### PR TITLE
Converting to use kotless

### PR DESCRIPTION
**Pull request by neilpower**
Converting to use kotless rather than deploying to EC2. 

Changed the datastore to use dynamodb

### CODE REVIEW

File: backend/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+plugins {
+    id("maven")
+    id("java")
+    id("io.spring.dependency-management") version "1.0.8.RELEASE"
+    id("org.jetbrains.kotlin.jvm")
+    id("application")
+    id("distribution")
+    id("com.github.johnrengelman.shadow")
+    id("io.kotless") version "0.1.5" apply true
```
Comment (benhergert) [inline backend/build.gradle.kts:11]: How stable is this version?

Comment (benhergert) [reply backend/build.gradle.kts:11]: Can you use the `kotlessVersion` variable declared further down in here?

File: backend/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        bucket = "hermesapp-test"
+
+        terraform {
+            profile = "default"
+            region = "us-west-2"
+        }
+    }
+
+    webapp {
+        // Optional parameter, by default technical name will be generated
```
Comment (benhergert) [inline backend/build.gradle.kts]: You can probably get rid of the commented out part

File: backend/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    implementation("com.github.kittinunf.fuel:fuel:$fuelVersion")
+    implementation("com.github.kittinunf.fuel:fuel-gson:$fuelVersion")
+
+    testImplementation("org.jetbrains.kotlin:kotlin-test")
+    testImplementation("org.jetbrains.kotlin:kotlin-test-junit")
+}
+
+kotless {
+    config {
+        bucket = "hermesapp-test"
```
Comment (benhergert) [inline backend/build.gradle.kts]: Is the `-test` postfix intentional?

File: backend/src/main/kotlin/com/hootsuite/hermes/Main.kt [hunk 4]
```diff
@@ -5,308 +5,150 @@ import com.github.kittinunf.fuel.gson.responseObject
...
+                    file("install.html")
+                    file("success.html")
+                    file("error.html")
+                    resource("favicon.ico")
+                }
 
-            //Reviews
-            get(Config.Endpoint.REVIEWS) { reviewsGet(call) }
+                // Webhooks from Github
+                post("/webhook") { webhookPost(call) }
```
Comment (benhergert) [inline backend/src/main/kotlin/com/hootsuite/hermes/Main.kt]: This is a Github webhook while the `/slack` endpoint is a Slack webhook. I'd reconsider the name here to something like `/github` that makes it clearer what kind of webhook this is.

File: backend/src/main/kotlin/com/hootsuite/hermes/Main.kt [hunk 5]
```diff
@@ -5,308 +5,150 @@ import com.github.kittinunf.fuel.gson.responseObject
...
-        filteredRequests.joinToString("<br>") { "<a href=\"${it.htmlUrl}\">$it</a>" }
+        // TODO Handle non-200
+        if (response.statusCode == HttpStatusCode.OK.value) {
+            val (slackAuth, _) = result
+            slackAuth?.incomingWebhook?.let { webhook ->
+                DatabaseUtils.createOrUpdateTeam(Team(webhook.channel, webhook.url))
+                if (webhook.channel == Config.ADMIN_CHANNEL) {
+                    logger.debug("Admin Channel registered")
+                }
+                // TODO Remove the /1 here and do full relative
```
Comment (benhergert) [inline backend/src/main/kotlin/com/hootsuite/hermes/Main.kt:114]: Why are you using the `/1/...` here?

File: backend/src/main/kotlin/com/hootsuite/hermes/github/GithubEventHandler.kt [hunk 6]
```diff
@@ -193,7 +192,7 @@ object GithubEventHandler {
      */
     fun ping(pingEvent: PingEvent) {
         val extraEvents = pingEvent.hook.events
-            .filterNot { it in SupportedEvents.values().map { it.eventName } }
+            .filterNot { event -> event in SupportedEvents.values().map { it.eventName } }
```
Comment (benhergert) [inline backend/src/main/kotlin/com/hootsuite/hermes/github/GithubEventHandler.kt:195]: 👍 for clarity

File: backend/src/main/kotlin/com/hootsuite/hermes/slack/SlashCommandHandler.kt [hunk 7]
```diff
@@ -26,7 +26,7 @@ object SlashCommandHandler {
                 }
                 if (parameters.size == 1) {
                     DatabaseUtils.createOrUpdateUserByGithubName(User(parameters.first(), slashCommand.username, team))
-                    if (DatabaseUtils.getTeamOrNull(team) != null) {
+                    if (DatabaseUtils.getTeamOrNull(team)) {
```
Comment (benhergert) [inline backend/src/main/kotlin/com/hootsuite/hermes/slack/SlashCommandHandler.kt]: Is this a new language feature I don't know about?

File: backend/src/main/tf/extensions.tf [hunk 8]
```diff
@@ -0,0 +1,52 @@
+resource "aws_dynamodb_table" "users" {
```
Comment (benhergert) [inline backend/src/main/tf/extensions.tf]: Not sure about general best practices, but `backend/src/main/tf/`isn't the first place I'd look for Terraform files.
I'd either pull it one level up under `backend/tf` or add a `tf` directory in the root of the project.

**Review by benhergert — COMMENTED** [review]
Why do you have all the dependencies embedded in the project under /lib? Can't you just pull them as a Maven dependency?

File: backend/build.gradle.kts (hunk 2 above)
Comment (neilpower) [reply backend/build.gradle.kts]: will do, I'll add back when I set up route53

File: backend/build.gradle.kts (hunk 1 above)
Comment (neilpower) [reply backend/build.gradle.kts:11]: no, sadly you need to use string literals in this context https://stackoverflow.com/questions/47704665/cant-use-project-extra-properties-in-plugin-block

File: backend/build.gradle.kts (hunk 2 above)
Comment (neilpower) [reply backend/build.gradle.kts]: done, will readd once I get a domain sorted

File: backend/build.gradle.kts (hunk 3 above)
Comment (neilpower) [reply backend/build.gradle.kts]: nope, fixed

File: backend/src/main/kotlin/com/hootsuite/hermes/Main.kt (hunk 4 above)
Comment (neilpower) [reply backend/src/main/kotlin/com/hootsuite/hermes/Main.kt]: good call, updated

File: backend/src/main/kotlin/com/hootsuite/hermes/Main.kt (hunk 5 above)
Comment (neilpower) [reply backend/src/main/kotlin/com/hootsuite/hermes/Main.kt:114]: it's an artifact of the API Gateway deployment

File: backend/src/main/kotlin/com/hootsuite/hermes/slack/SlashCommandHandler.kt (hunk 7 above)
Comment (neilpower) [reply backend/src/main/kotlin/com/hootsuite/hermes/slack/SlashCommandHandler.kt]: changed the method signature without updating the name of the method, updated

File: backend/src/main/tf/extensions.tf (hunk 8 above)
Comment (neilpower) [reply backend/src/main/tf/extensions.tf]: agreed, moved to the module root

**Pull request merged**